UefiPayloadPkg/Library/FdtParserLib/FdtParserLib.c

```diff
@@ -0,0 +1,699 @@
+
+        if (AsciiStrnCmp (NodePtr->Name, GmaStr, AsciiStrLen (GmaStr)) == 0) {
+          DEBUG ((DEBUG_INFO, "  Found gma@ node \n"));
+          //
+          // Build Graphic info HOB .
+          //
+          GraphicsDev = BuildGuidHob (&gEfiGraphicsDeviceInfoHobGuid, sizeof (EFI_PEI_GRAPHICS_DEVICE_INFO_HOB));
+          ASSERT (GraphicsDev != NULL);
+          if (GraphicsDev == NULL) {
+            break;
+          }
+
+          SetMem (GraphicsDev, sizeof (EFI_PEI_GRAPHICS_DEVICE_INFO_HOB), 0xFF);
+
+          PropertyPtr = FdtGetProperty (Fdt, SubNode, "vendor-id", &TempLen);
+          ASSERT (TempLen > 0);
+          if (TempLen > 0) {
+            Data32 = (UINT32 *)(PropertyPtr->Data);
+            GmaID  = (UINT16)Fdt32ToCpu (*Data32);
+            DEBUG ((DEBUG_INFO, "\n   vendor-id"));
+            DEBUG ((DEBUG_INFO, "  %016lX\n", GmaID));
+
+            GraphicsDev->VendorId = GmaID;
+          }
+
+          PropertyPtr = FdtGetProperty (Fdt, SubNode, "device-id", &TempLen);
+          ASSERT (TempLen > 0);
+          if (TempLen > 0) {
+            Data32 = (UINT32 *)(PropertyPtr->Data);
+            GmaID  = (UINT16)Fdt32ToCpu (*Data32);
+            DEBUG ((DEBUG_INFO, "\n   device-id"));
+            DEBUG ((DEBUG_INFO, "  %016lX\n", GmaID));
+
+            GraphicsDev->DeviceId = GmaID;
+          }
+
+          PropertyPtr = FdtGetProperty (Fdt, SubNode, "revision-id", &TempLen);
+          ASSERT (TempLen > 0);
+          if (TempLen > 0) {
+            Data32 = (UINT32 *)(PropertyPtr->Data);
+            GmaID  = (UINT16)Fdt32ToCpu (*Data32);
+            DEBUG ((DEBUG_INFO, "\n   revision-id"));
+            DEBUG ((DEBUG_INFO, "  %016lX\n", GmaID));
+
+            GraphicsDev->RevisionId = (UINT8)GmaID;
+          }
+
+          PropertyPtr = FdtGetProperty (Fdt, SubNode, "subsystem-vendor-id", &TempLen);
+          ASSERT (TempLen > 0);
+          if (TempLen > 0) {
+            Data32 = (UINT32 *)(PropertyPtr->Data);
+            GmaID  = (UINT16)Fdt32ToCpu (*Data32);
+            DEBUG ((DEBUG_INFO, "\n   subsystem-vendor-id"));
+            DEBUG ((DEBUG_INFO, "  %016lX\n", GmaID));
+
+            GraphicsDev->SubsystemVendorId = GmaID;
+          }
+
+          PropertyPtr = FdtGetProperty (Fdt, SubNode, "subsystem-id", &TempLen);
+          ASSERT (TempLen > 0);
+          if (TempLen > 0) {
+            Data32 = (UINT32 *)(PropertyPtr->Data);
+            GmaID  = (UINT16)Fdt32ToCpu (*Data32);
+            DEBUG ((DEBUG_INFO, "\n   subsystem-id"));
+            DEBUG ((DEBUG_INFO, "  %016lX\n", GmaID));
+
+            GraphicsDev->SubsystemId = GmaID;
+          }
+        }
+
+        if (AsciiStrnCmp (NodePtr->Name, "espi", AsciiStrLen ("espi")) == 0) {
+          SSubNode = FdtFirstSubnode (Fdt, SubNode); // serial
+          //
+          // Create SerialPortInfo HOB.
+          //
+          Serial = BuildGuidHob (&gUniversalPayloadSerialPortInfoGuid, sizeof (UNIVERSAL_PAYLOAD_SERIAL_PORT_INFO));

```

Comment:
this is for serial under espi configuration, but we also have serial device under PCI configuration case, please also add such support.

Reply:
resolved